### PR TITLE
Moving to DUB and latest DMD

### PR DESCRIPTION
**Pull request by NCrashed**
It would be much easily to use cl4d if there is dub support. 

Notes:
- Dub package description. DerelictSDL2 and DerelictGL3 are moved to package dependencies.
- Sources are moved to source subfolder.
- Compilation fixes for dmd v2.065.
- Auto linking to OpenCL libraries.
- Recently I had problems with `CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR` error, that wasn't caught by old cl4d version.
- Examples are moved to new versions of dependencies.
- README is now markdown to handle code snippets.
- Very difficult to disable debug output of dependencies while using dub, therefore debug `writef`s are moved to special version `CL4D_VERBOSE`
- Trailing `\0` wasn't deleted from `getInfoString` method, that cause build log hiding while printing to console.

Tested on Fedora 20 x86_64 and Win7 x86

### CODE REVIEW

**Comment by Trass3r** [comment]
Thanks for your contribution!
Is there any value in moving all the files? We have a dub json anyway.

**Comment by NCrashed** [comment]
> Is there any value in moving all the files?

Just for standardization, dub packages are usually have dedicated folder for sources.

File: source/CLGLInterop.d [hunk 1]
```diff
@@ -48,22 +49,33 @@ CLCommandQueue CQ;
...
+	DerelictSDL2.load();
+	
 	// create a window plus OpenGL context
-	DerelictSDL.load();
-	SDL_Init(SDL_INIT_VIDEO | SDL_INIT_EVENTTHREAD);
-
-	SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
-	SDL_Surface* screen = SDL_SetVideoMode(1024, 768, 0, SDL_OPENGL|SDL_RESIZABLE|SDL_HWSURFACE);
-
+	assert(SDL_Init(SDL_INIT_VIDEO) == 0, "Error while initing SDL!");
```
Comment (Trass3r) [inline source/CLGLInterop.d]: side effect in assert

File: source/CLGLInterop.d [hunk 2]
```diff
@@ -85,14 +97,23 @@ void main()
...
-					{} // Uh oh, we couldn't set the new video mode??
-					break;
-
+				case SDL_WINDOWEVENT:
+				{
+				    switch(evt.window.event)
+				    {
+				        case SDL_WINDOWEVENT_RESIZED:
+				        {
+        					//screen = SDL_SetVideoMode(evt.resize.w, evt.resize.h, 0, SDL_OPENGL|SDL_RESIZABLE|SDL_HWSURFACE);
```
Comment (Trass3r) [inline source/CLGLInterop.d]: remnant?

File: source/opencl/c/cl.d [hunk 3]
```diff
@@ -29,8 +29,8 @@ package string bringToCurrentScope(alias EnumType)()
 
 extern(System):
 
-//! these MUST be typedefs
-typedef const(void*)
+//! these MUST be aliass
```
Comment (Trass3r) [inline source/opencl/c/cl.d]: Now the comments make no sense anymore

File: source/CLGLInterop.d (hunk 1 above)
Comment (NCrashed) [inline source/CLGLInterop.d]: will replace to enforce

File: source/CLGLInterop.d (hunk 2 above)
Comment (NCrashed) [inline source/CLGLInterop.d]: forgot that, will remove

File: source/opencl/c/cl.d (hunk 3 above)
Comment (NCrashed) [inline source/opencl/c/cl.d]: By the way, I've run into the pitfall later when `T.stringof` was outputing `const(void*)`. Found a workaround below: usage of explicit `TName` enum.

P.S. I tried library based typedef, but `T.stringof` returned wrapper name.

**Comment by Trass3r** [comment]
Even then I'd prefer src over source. But I don't really like a useless folder in a folder. And I want to rename the package to cl4d. I pushed a test branch where I'm experimenting.

**Comment by NCrashed** [comment]
Removing useless folder is ok, I have no objections about it. It's a good time for renaming, there isn't so much code to be broken for now.

**Comment by Trass3r** [comment]
Also tried switching to DerelictCL (905c95564c) but unfortunately it lacks the enum types and also some function pointer types.
But it also brought up another issue. How to deal with const-correctness?
Any C function might modify the passed opaque object.

**Comment by NCrashed** [comment]
I think, const should be inferred from kronos documentation, and if standard says nothing, then the only thing one can do is don't use const with these functions.

P.S. Anyway const for opaque structures is only user side information, it is reference that WE cannot change, but cl dirver can.

**Comment by Trass3r** [comment]
Sure. I think my solution with const(void*) is the best compromise. It prevents changing the pointer you received. And you can tag cl4d methods as const where it makes sense like the getInfo ones without having to go through all C functions and adding const to some (which is only a guess as well).

**Comment by Trass3r** [comment]
Has long been integrated into master.

**Pull request closed without merging**